torchbox/wagtail-torchbox
# New StreamField-powered SubPropositionPage based on SubServicePage

[Link to Ticket](https://torchbox.monday.com/boards/1192293412/pulses/1193045284)

### Description of Changes Made

#### Overview

This PR introduces a new page type — `SubPropositionPage`, based on the existing `SubServicePage`, which has a fixed ordering of content on the rendered page. The  `SubPropositionPage` reorganizes the page content to use Streamfield blocks, thereby allowing content editors the freedom to arrange the content in any order.

#### More details

- Sub-service pages have 5 major sections:
  - Key Points
  - Testimonials
  - Processes
  - Work (Case Studies)
  - Thinking (Featured Blog Posts)
- For each of the above sections, I have created a new `StructBlock` with its own template
-  A new `StreamBlock` puts all the above blocks together, also including the blocks in `tbx.core.blocks.PageSectionStoryBlock`, which is part of the existing `SubServicePage`. 
  - these blocks are distinguished by grouping them into two groups: **Page sections** and **Content blocks**
  - **Content blocks** are 'standard' StreamField blocks which are enclosed in a `div` with class `streamfield`. However, when you add the **Page sections** blocks inside the same div, the styling breaks. Therefore, I created a new template for the new `StreamBlock`, with logic to ensure we don't place **Page sections** inside `<div class="streamfield">`. See 4fd8146a for details.
- The existing `patterns/molecules/hero/hero.html` template has an include for `patterns/molecules/in-page-nav/in-page-nav.html`.
  - A variable named `item` is passed to the included template, with the value of the `page` variable from the current context assigned to it. This `page` is expected to have certain fields/attributes, such as `process_section_title`, `testimonials_section_title`, etc. However, our new `SubPropositionPage` has been stripped of many of these fields. Moreover, the ordering of the fields in the in-page nav template is fixed, which would not work well for our use-case, where content editors can freely order the page sections.
  - Therefore, I have created new templates `patterns/molecules/hero/hero--sub-proposition.html` and `patterns/molecules/in-page-nav/in-page-nav--sub-proposition.html` to solve the above problem. See 52d66ac8 for details.
- I have made small changes (1c9b1234) to `patterns/organisms/client-block/client-block.html` and `patterns/organisms/latest-posts/latest-posts.html` to handle two situations where certain `page` attributes are not available. It seemed more appropriate in this case to just modify the templates to include this logic than to create new templates.
- I have updated the stylesheet `tbx/static_src/sass/components/_page.scss` to remove padding from the `template__sub-proposition-page` class, as is the case for `template__sub-service-page` and other templates (See eead1d50 for details).
- I have written a few tests for the new `SubPropositionPage`. I had to create several factories (and tests for those factories) in order to create a more complete factory for the new page. I could have added more tests, but in order to avoid making this PR larger than it needs to be, it'd be best to add more tests in the next PR, that will focus on migrating data from existing `SubServicePage`s to the new `SubPropositionPage`, as indicated in the note below.
- `tbx.core.utils.migrations.for_each_page_revision` has been breaking, as a result of a `FieldError: Cannot resolve keyword 'page' into field`. I have fixed this in fd2a20d6.

> **Note**
>
> - The `SubServicePage` will be deprecated in a separate PR, alongside the `ServicePage` model, which is replaced by the `PropositionPage`.
> - A separate PR will be created to migrate data from existing `SubServicePage`s to the new `SubPropositionPage`.

### How to Test

-  Choose any existing **Sub service page** to use as a basis for the new **Sub proposition page** that you will be creating in the next step.
- In Wagtail Admin, create a new **Sub proposition page** under any of the existing **Proposition pages** (You should not be able to create it anywhere else).
- Create content for the new page using the selected sub service page's content, so that you essentially have a clone of the sub service page. You will observe that fields for the 5 page sections on the sub service page (Key Points, Testimonials, Processes, Work, Thinking) are now part of the **content** stream field on the new page.
- Check that you can save the Page without errors. 
- Check that the new page looks exactly the same as the sub service page.
- Check that the in-page navigation woks correctly, and that the order of the navigation items matches the order of the sections on the page
- Now, you can 
  - reorder the streamfield blocks for any of the page sections, and check that they are rendered correctly on the front end. Also verify that the in-page navigation is updated accordingly, and isn't broken,
  - remove one or two of the streamfield blocks for any of the page sections, and repeat the above checks,
  - add other streamfield blocks (from the **Content block** group, not the **Page section** group) as you please, and check that they are rendered correctly (at the right position, with the expected styling, maintaining the expected behaviour) and do not break anything else on the page.

### Screenshots

<details>
<summary>Before (Wagtail Admin)</summary>
    
![admin](https://github.com/torchbox/wagtail-torchbox/assets/7713776/e3626f3f-1d5b-48c7-babb-e2e11287afcc)
    
</details>

<details>
<summary>Before (Template)</summary>

![existing_frontend](https://github.com/torchbox/wagtail-torchbox/assets/7713776/7b3009dd-9483-482d-adda-5d4256d605fd)
    
</details>

<details>
    <summary>After (Wagtail Admin)</summary>

 ![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/d6e4f574-0805-49c3-81f1-70fd2ae508be)
 
</details>

<details>
    <summary>After (Template)</summary>
   
![new_frontend](https://github.com/torchbox/wagtail-torchbox/assets/7713776/04c45200-2117-4c95-9740-040905128cfb)

</details>

### MR Checklist

  *The changes do not need any dedicated docs. However, I have added comments in the code where appropriate, and documented all the changes in this PR*
